wallet: bugfix, invalid crypted key "checksum_valid" set

furszy opened the pull request:
At wallet load time, the crypted key "checksum_valid" variable is always set to false. Which, on every wallet decryption call, forces the process to re-write all the ckeys to db when it's not needed.

Note:
The first commit fixes the issue, the two commits in the middle are cleanups so `DuplicateMockDatabase`
can be used without duplicating code. And, the last one is pure test coverage for the crypted keys loading
process.

Includes test coverage for the following scenarios:

1) "All ckeys checksums valid" test:
Loads an encrypted wallet with all the crypted keys with a valid checksum and
verifies that 'CWallet::Unlock' doesn't force an entire crypted keys re-write.

    (we force a complete ckeys re-write if we find any missing crypted key checksum
during the wallet loading process)

2) "Missing checksum in one ckey" test:
Verifies that loading up a wallet with, at least one, 'ckey' with no checksum
triggers a complete re-write of the crypted keys.

3) "Invalid ckey checksum error" test:
Verifies that loading up a ckey with an invalid checksum stops the wallet loading
process with a corruption error.

4) "Invalid ckey pubkey error" test:
Verifies that loading up a ckey with an invalid pubkey stops the wallet loading
process with a corruption error.

furszy committed:
refactor: unify test/util/wallet.h with wallet/test/util.h

files share the same purpose, and we shouldn't have wallet code
inside the test directory.

This later is needed to use wallet util functions in the bench
and test binaries without be forced to duplicate them.

furszy committed:
refactor: move DuplicateMockDatabase to wallet/test/util.h

furszy committed:
wallet: bugfix, invalid crypted key "checksum_valid" set

At wallet load time, we set the crypted key "checksum_valid" variable always to false.
Which, on every wallet decryption call, forces the process to re-write the entire ckeys to db when
it's not needed.

achow101 commented:
Good catch! Do you think it's possible to have a test for this?

furszy committed:
test: load wallet, coverage for crypted keys

Adds test coverage for the wallet's crypted key loading from db process.
The following scenarios are covered:

1) "All ckeys checksums valid" test:
   Loads an encrypted wallet with all the crypted keys with a valid checksum and
   verifies that 'CWallet::Unlock' doesn't force an entire crypted keys re-write.

   (we force a complete ckeys re-write if we find any missing crypted key checksum
    during the wallet loading process)

2) "Missing checksum in one ckey" test:
   Verifies that loading up a wallet with, at least one, 'ckey' with no checksum
   triggers a complete re-write of the crypted keys.

3) "Invalid ckey checksum error" test:
   Verifies that loading up a ckey with an invalid checksum stops the wallet loading
   process with a corruption error.

4) "Invalid ckey pubkey error" test:
   Verifies that loading up a ckey with an invalid pubkey stops the wallet loading
   process with a corruption error.

Tweakwild113 commented:
Wat to do

furszy commented:
> Good catch! Do you think it's possible to have a test for this?

Yep :), added test coverage for it and more inside d9dd212d.

#### Covered the following scenarios

1) "All ckeys checksums valid" test:
   Loads an encrypted wallet with all the crypted keys with a valid checksum and
   verifies that 'CWallet::Unlock' doesn't force an entire crypted keys re-write.

   (we force a complete ckeys re-write if we find any missing crypted key checksum
    during the wallet loading process)

2) "Missing checksum in one ckey" test:
   Verifies that loading up a wallet with, at least one, 'ckey' with no checksum
   triggers a complete re-write of the crypted keys.

3) "Invalid ckey checksum error" test:
   Verifies that loading up a ckey with an invalid checksum stops the wallet loading
   process with a corruption error.

4) "Invalid ckey pubkey error" test:
   Verifies that loading up a ckey with an invalid pubkey stops the wallet loading
   process with a corruption error.

Extra note:
The two commits in the middle are cleanups so I can use the `DuplicateMockDatabase`
function without duplicating code.

achow101 commented on src/Makefile.test_util.include:
```diff
@@ -19,7 +19,7 @@ TEST_UTIL_H = \
   test/util/txmempool.h \
   test/util/validation.h \
-  test/util/wallet.h
+  wallet/test/util.h
```
In cd62eddbcfba652a238374d62f70181cb3ad9075 "refactor: unify test/util/wallet.h with wallet/test/util.h"

nit: ISTM these should be guarded by `ENABLE_WALLET`? They aren't used by non-wallet benchmarks or tests, so it doesn't break anything.

This would also remove the need to have `#ifdef ENABLE_WALLET` in those files too.

furszy replied:
hmm yeah, could also remove them from `BITCOIN_TEST_SUITE` as `TEST_UTIL_H` is already included there.

achow101 commented on src/wallet/test/util.cpp:
```diff
@@ -11,10 +11,11 @@
-
 #include <memory>
 
+const std::string ADDRESS_BCRT1_UNSPENDABLE = "bcrt1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq3xueyj";
```
In cd62eddbcfba652a238374d62f70181cb3ad9075 "refactor: unify test/util/wallet.h with wallet/test/util.h"

nit: Although this were originally in `src/test/util/wallet.h`, it doesn't really make sense to be in the wallet's utils since it is not a wallet specific address. Could probably go in `src/test/util/script.h` or `src/test/util/transaction_utils.h`

achow101 reviewed (commented):
ACK 4db80c90b02b70a71a3ac1d614d779d222561525

furszy commented:
Updated per feedback, thanks achow101.

[Small Diff](https://github.com/bitcoin/bitcoin/compare/4db80c90b02b70a71a3ac1d614d779d222561525..13d97608297bd56ed033d0e754d2e50447b02af0):
1) Added `ENABLE_WALLET` guard for the `wallet/test/util.h` and util.cpp files at the test_util makefile level.
2) Moved `ADDRESS_BCRT1_UNSPENDABLE` from `wallet/test/util.h` to `wallet_balance.cpp` (it's only used there).

aureleoules reviewed (approved):
ACK 13d97608297bd56ed033d0e754d2e50447b02af0

Cherry-picked the last three commits on master and verified the tests do not pass. I also verified that the tests correctly test the behavior.

achow101 commented:
ACK 13d97608297bd56ed033d0e754d2e50447b02af0

furszy commented on src/wallet/test/walletload_tests.cpp:
```diff
@@ -50,5 +51,139 @@ BOOST_FIXTURE_TEST_CASE(wallet_load_unknown_descriptor, TestingSetup)
+
+    {   // Context setup.
+        // Create and encrypt legacy wallet
+        std::shared_ptr<CWallet> wallet(new CWallet(m_node.chain.get(), "", m_args, CreateMockWalletDatabase()));
```
k thanks, will do if have to retouch.

furszy commented on src/wallet/test/walletload_tests.cpp:
```diff
@@ -50,5 +51,139 @@ BOOST_FIXTURE_TEST_CASE(wallet_load_unknown_descriptor, TestingSetup)
     }
 }
 
+bool HasAnyRecordOfType(WalletDatabase& db, const std::string& key)
```
thanks, will do if have to retouch

luke-jr reviewed (approved):
utACK cc5a5e81217506ec6f9fff34056290f8f40a7396 by itself (did not review refactors or new test)
